query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/QueryBuilderDsl.java

```diff
@@ -0,0 +1,742 @@
+  }
+
+  /** Builds a relation testing a token generated from a set of columns. */
+  static TokenRelationBuilder isTokenFromIds(Iterable<CqlIdentifier> identifiers) {
+    return new DefaultTokenRelationBuilder(identifiers);
+  }
+
+  /** Var-arg equivalent of {@link QueryBuilderDsl#isTokenFromIds(Iterable)}. */
+  static TokenRelationBuilder isToken(CqlIdentifier... identifiers) {
+    return isTokenFromIds(Arrays.asList(identifiers));
+  }
+
+  /**
+   * Equivalent of {@link QueryBuilderDsl#isTokenFromIds(Iterable)} with raw strings; the names are
+   * converted with {@link CqlIdentifier#fromCql(String)}.
+   */
+  static TokenRelationBuilder isToken(Iterable<String> names) {
+    return isTokenFromIds(Iterables.transform(names, CqlIdentifier::fromCql));
+  }
+
+  /** Var-arg equivalent of {@link #isToken(Iterable)}. */
+  static TokenRelationBuilder isToken(String... names) {
+    return isToken(Arrays.asList(names));
+  }
+
+  /** Builds a relation testing a set of columns, as in {@code WHERE (c1, c2, c3) IN ...}. */
+  static TupleRelationBuilder isTupleOfIds(Iterable<CqlIdentifier> identifiers) {
+    return new DefaultTupleRelationBuilder(identifiers);
+  }
+
+  /** Var-arg equivalent of {@link #isTupleOfIds(Iterable)}. */
+  static TupleRelationBuilder isTuple(CqlIdentifier... identifiers) {
+    return isTupleOfIds(Arrays.asList(identifiers));
+  }
+
+  /**
+   * Equivalent of {@link #isTupleOfIds(Iterable)} with raw strings; the names are converted with
+   * {@link CqlIdentifier#fromCql(String)}.
+   */
+  static TupleRelationBuilder isTuple(Iterable<String> names) {
+    return isTupleOfIds(Iterables.transform(names, CqlIdentifier::fromCql));
+  }
+
+  /** Var-arg equivalent of {@link #isTuple(Iterable)}. */
+  static TupleRelationBuilder isTuple(String... names) {
+    return isTuple(Arrays.asList(names));
+  }
+
+  /** Builds a relation on a custom index. */
+  static Relation isCustomIndex(CqlIdentifier indexId, Term expression) {
+    return new CustomIndexRelation(indexId, expression);
+  }
+
+  /**
+   * Shortcut for {@link #isCustomIndex(CqlIdentifier, Term)
+   * isCustomIndex(CqlIdentifier.fromCql(indexName), expression)}
+   */
+  static Relation isCustomIndex(String indexName, Term expression) {
+    return isCustomIndex(CqlIdentifier.fromCql(indexName), expression);
+  }
+
+  /**
+   * An ordered set of anonymous terms, as in {@code WHERE (a, b) = (1, 2)} (on the right-hand
+   * side).
+   *
+   * <p>For example, this can be used for the right-hand side of {@link
+   * QueryBuilderDsl#isTuple(String...)}.
+   */
+  static Term tuple(Iterable<? extends Term> components) {
+    return new TupleTerm(components);
+  }
+
+  /** Var-arg equivalent of {@link #tuple(Iterable)}. */
+  static Term tuple(Term... components) {
+    return tuple(Arrays.asList(components));
+  }
+
+  /** The sum of two terms, as in {@code WHERE k = left + right}. */
+  static Term sum(Term left, Term right) {
```

Comment:
I like the idea of `Term` in general, forcing users to be more expressive about what they are passing (i.e. `literal(v, codec)` and `raw(string)` differentiating passing in a value or the literal value string) and also pointing users more in the direction of using `bindMarker` instead of passing in explicit values.

I can see some initial frustration about not being allowed to passing actual values, but I agree that it's probably for the best in the long run.

Reply:
Yes, I think we've been overestimating the need for actual values a bit, in most cases they should be handled with bind markers. And for the remaining cases `literal()` is not that much of a stretch.

Reply:
I think the one big win of being able to pass in names and values as is is that it makes the resulting code much simpler.  I.E., compare this from the tests:

```java
selectFrom("foo").function("f", getColumn("c1"), getSum(getColumn("c2"), raw("1")));
```

to:

```java
selectFrom("foo").function("f", "c1", getSum("c2", "1")));
```

The second implementation is much easier to form and understand in my opinion.

From a user's perspective, I'm almost always passing in column names and values as I expect them to be interpreted as in the query text, so why do I have to be so explicit about that?

On the other hand I will agree that being explicit does have a lot of value.  It requires the user to express intent so the driver can always make the right decision, but there is also a cost in that it makes the query builder a little more difficult to work with and a little more weighty.  I'm a bit conflicted with regards to which is the best approach.

Reply:
I tried to provide overloads for the most straightforward cases, for example:

```java
getElement(Selector, Term)
getElement(CqlIdentifier, Term)
getElement(String, Term)
```

But each time we add a new alternative it's a cartesian product, in the end we can't have 40 overloads of every method, so we need to find the right balance. Java's lack of syntactic sugar (optional parameters, implicits...) really stings for this kind of API.

But also, I don't think we should focus primarily on how "nice" the code looks. The primary goal of the query builder is to provide a means to generate queries programmatically, not an alternative to raw strings for static queries (I have a paragraph about that  in the upcoming manual, hope it will make my point more clear). In code that truly generates expressions dynamically, you *need* all those factory methods.

Reply:
One more thing that came to me is that the more we give up on typing, the more ambiguous things become. In your example:

```java
selectFrom("foo").function("f", "c1", getSum("c2", "1")));
```

How do we know if we're dealing with `f(c1,c2+1)` or `f.c1(c2+1)`?

Internally, that also translates as methods that receive `Object` and a bunch of `instance of` check, like we have in driver 3 (I'd like to avoid repeating that).